src/workerd/util/sqlite.c++

```diff
@@ -0,0 +1,835 @@
+    int _ec = code; \
+    KJ_ASSERT(_ec == SQLITE_OK, "SQLite failed: " #code, sqlite3_errstr(_ec), ##__VA_ARGS__); \
+  } while (false)
+// Make a SQLite call and check the returned error code. Use this version when the call is not
+// associated with an open DB connection.
+
+#define SQLITE_CALL(code, ...) do { \
+    int _ec = code; \
+    /* SQLITE_MISUSE doesn't put error info on the database object, so check it separately */ \
+    KJ_ASSERT(_ec != SQLITE_MISUSE, "SQLite misused: " #code, ##__VA_ARGS__); \
+    KJ_ASSERT(_ec == SQLITE_OK, "SQLite failed: " #code, sqlite3_errmsg(db), ##__VA_ARGS__); \
+  } while (false)
+// This version requires the scope to contain a variable named `db` which is of type sqlite3*, or
+// can convert to it.
+
+#define SQLITE_CALL_FAILED(code, error, ...) do { \
+    KJ_ASSERT(error != SQLITE_MISUSE, "SQLite misused: " code, ##__VA_ARGS__); \
+    KJ_ASSERT(error == SQLITE_OK, "SQLite failed: " code, sqlite3_errmsg(db), ##__VA_ARGS__); \
+  } while (false);
+// Version of `SQLITE_CALL` that can be called after inspecting the error code, in case some codes
+// aren't really errors.
+
+namespace {
+
+void disposeSqlite(sqlite3_stmt* stmt) {
+  sqlite3_finalize(stmt);
+
+  // Note that any returned error code is actually the last error to occur while executing the
+  // statement. This does not really mean that finalization failed, and the error in question
+  // should have been checked and reported earlier. So, we ignore it here.
+}
+
+template <typename T>
+class SqliteDisposer: public kj::Disposer {
+public:
+  void disposeImpl(void* pointer) const override {
+    disposeSqlite(reinterpret_cast<T*>(pointer));
+  }
+};
+
+template <typename T>
+kj::Own<T> ownSqlite(T* obj) {
+  static const SqliteDisposer<T> disposer;
+  return kj::Own<T>(obj, disposer);
+}
+
+}  // namespace
+
+// =======================================================================================
+
+Sqlite::Sqlite(Vfs& vfs, kj::PathPtr path) {
+  SQLITE_CALL_NODB(sqlite3_open_v2(path.toString().cStr(), &db,
+                                   SQLITE_OPEN_READONLY, vfs.getName().cStr()));
+}
+
+Sqlite::Sqlite(Vfs& vfs, kj::PathPtr path, kj::WriteMode mode) {
+  int flags = SQLITE_OPEN_READWRITE;
+  if (kj::has(mode, kj::WriteMode::CREATE)) {
+    flags |= SQLITE_OPEN_CREATE;
+
+    if (kj::has(mode, kj::WriteMode::CREATE_PARENT) && path.size() > 1) {
+      // SQLite isn't going to try to create the parent directory so let's try to create it now.
+      vfs.directory.openSubdir(path.parent(),
+          kj::WriteMode::CREATE | kj::WriteMode::MODIFY | kj::WriteMode::CREATE_PARENT);
+    }
+  }
+  KJ_REQUIRE(kj::has(mode, kj::WriteMode::MODIFY), "SQLite doesn't support create-exclusive mode");
+
+  SQLITE_CALL_NODB(sqlite3_open_v2(path.toString().cStr(), &db,
+                                   flags, vfs.getName().cStr()));
+}
+
+Sqlite::~Sqlite() noexcept(false) {
+  auto err = sqlite3_close(db);
+  if (err == SQLITE_BUSY) {
```

Comment:
I'm not intending to allow dependent objects to outlive the database -- that's still an error. I just wrote this code so that instead of crashing or leaking memory, we log an error and escape gracefully.